### PR TITLE
Actually test GetBytes and GetChars in System.Text.Encoding

### PR DESCRIPTION
**Pull request by hughbe**
- Previous tests for GetBytes and GetChars did not test anything apart
from the return value: the method could've created any byte[] or char[]
result and we wouldn't have verified it.
- Also adds more coverage by testing more overloads of GetBytes and
GetChars

### CODE REVIEW

File: src/System.Text.Encoding/tests/EncodingTestHelpers.cs [hunk 1]
```diff
@@ -23,17 +23,53 @@ public static void GetByteCount(Encoding encoding, string chars, int index, int
...
+                // Bytes outside the range should be ignored
+                Assert.Equal(originalBytes[i], bytes[i]);
+            }
+            for (int i = byteIndex; i < byteIndex + byteCount; i++)
+            {
+                Assert.Equal(expectedBytes[i - byteIndex], bytes[i]);
+            }
+            for (int i = byteIndex + byteCount; i < bytes.Length; i++)
+            {
+                // Chars outside the range should be ignored
```
Comment (stephentoub) [inline src/System.Text.Encoding/tests/EncodingTestHelpers.cs]: Nit: bytes?

File: src/System.Text.Encoding/tests/EncodingTestHelpers.cs [hunk 2]
```diff
@@ -47,10 +83,42 @@ public static void GetCharCount(Encoding encoding, byte[] bytes, int index, int
...
-            Assert.Equal(expected, result);
+            char[] originalChars = (char[])chars.Clone();
+
+            // Use GetChars(byte[])
+            if (byteIndex == 0 && byteCount == bytes.Length)
+            {
+                char[] resultBasic = encoding.GetChars(bytes);
+                VerifyGetChars(resultBasic, 0, resultBasic.Length, originalChars, expectedChars);
+            }
+            // Use GetChars(byte[], int int)
```
Comment (stephentoub) [inline src/System.Text.Encoding/tests/EncodingTestHelpers.cs]: Nit: missing comma

File: src/System.Text.Encoding/tests/UnicodeEncoding/UnicodeEncodingGetBytes.cs [hunk 3]
```diff
@@ -11,20 +11,22 @@ public class UnicodeEncodingGetBytes
...
+            string testString = "Hello World";
+            byte[] testStringBytes = new byte[] { 72, 0, 101, 0, 108, 0, 108, 0, 111, 0, 32, 0, 87, 0, 111, 0, 114, 0, 108, 0, 100, 0 };
 
-            yield return new object[] { string.Empty, 0, 0, new byte[0], 0, 0 };
+            yield return new object[] { testString, 0, testString.Length, new byte[30], 5, testStringBytes };
+            yield return new object[] { testString, 0, 1, new byte[30], 0, new byte[] { 72, 0 } };
+            yield return new object[] { testString, 0, 0, new byte[20], 20, new byte[0] };
+            yield return new object[] { testString, testString.Length, 0, new byte[20], 20, new byte[0] };
+
+            yield return new object[] { string.Empty, 0, 0, new byte[0], 0, new byte[0] };
```
Comment (stephentoub) [inline src/System.Text.Encoding/tests/UnicodeEncoding/UnicodeEncodingGetBytes.cs]: I suggest adding an entry with all bytes, 0 through 255.

**Comment by stephentoub** [comment]
Some nits, otherwise LGTM.

**Comment by stephentoub** [comment]
@dotnet-bot test this please

File: src/System.Text.Encoding/tests/UnicodeEncoding/UnicodeEncodingGetBytes.cs (hunk 3 above)
Comment (hughbe) [inline src/System.Text.Encoding/tests/UnicodeEncoding/UnicodeEncodingGetBytes.cs]: fixed

File: src/System.Text.Encoding/tests/EncodingTestHelpers.cs (hunk 2 above)
Comment (hughbe) [inline src/System.Text.Encoding/tests/EncodingTestHelpers.cs]: fixed

File: src/System.Text.Encoding/tests/EncodingTestHelpers.cs (hunk 1 above)
Comment (hughbe) [inline src/System.Text.Encoding/tests/EncodingTestHelpers.cs]: fixed

File: src/System.Text.Encoding/tests/UnicodeEncoding/UnicodeEncodingGetBytes.cs [hunk 4]
```diff
@@ -19,7 +19,13 @@ public static IEnumerable<object[]> GetBytes_TestData()
             yield return new object[] { testString, 0, 0, new byte[20], 20, new byte[0] };
             yield return new object[] { testString, testString.Length, 0, new byte[20], 20, new byte[0] };
 
+            byte[] allBytes = new byte[byte.MaxValue];
+            for (byte i = 0; i < allBytes.Length; i++)
+            {
+                allBytes[i] = i;
+            }
```
Comment (stephentoub) [inline src/System.Text.Encoding/tests/UnicodeEncoding/UnicodeEncodingGetBytes.cs]: Isn't this missing the value 255?

File: src/System.Text.Encoding/tests/UnicodeEncoding/UnicodeEncodingGetBytes.cs [hunk 5]
```diff
@@ -11,20 +11,28 @@ public class UnicodeEncodingGetBytes
...
+            yield return new object[] { testString, 0, testString.Length, new byte[30], 5, testStringBytes };
+            yield return new object[] { testString, 0, 1, new byte[30], 0, new byte[] { 72, 0 } };
+            yield return new object[] { testString, 0, 0, new byte[20], 20, new byte[0] };
+            yield return new object[] { testString, testString.Length, 0, new byte[20], 20, new byte[0] };
+
+            byte[] allBytes = new byte[byte.MaxValue + 1];
+            for (byte i = 0; i < allBytes.Length; i++)
+            {
+                allBytes[i] = i;
+            }
```
Comment (stephentoub) [inline src/System.Text.Encoding/tests/UnicodeEncoding/UnicodeEncodingGetBytes.cs]: Isn't this now an infinite loop?

Comment (hughbe) [inline src/System.Text.Encoding/tests/UnicodeEncoding/UnicodeEncodingGetBytes.cs]: good spot. dunno how that got past me - I was too lazy to build locally :(

**Comment by hughbe** [comment]
Test Innerloop OSX Release Build and Test please

**Pull request merged**